### PR TITLE
feat(slack): Initial draft of Slack connector

### PR DESCRIPTION
**Pull request by dcramer**
Add a Slack connector to the new Integrations framework. For Slack it will captures the bot user and tokens, and is bound to a single team.

### CODE REVIEW

File: src/sentry/integrations/slack/integration.py
Comment (dcramer) [inline src/sentry/integrations/slack/integration.py]: not used yet, was mocking this out and forgot I left it in here..

right now Slack doesn't actually keep the association with the user, or their access token

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/integrations/oauth.py


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/conf/server.py
Comment (mitsuhiko) [inline src/sentry/conf/server.py]: is this built in now?

**Comment by mitsuhiko** [comment]
Looks good. Nothing that sounds like calling out.

File: src/sentry/conf/server.py
Comment (dcramer) [reply src/sentry/conf/server.py]: @mitsuhiko yeah i was using this as a way to test defaults, but we decided there was no point in maintaining first-party integrations in a different repo so just adding it to the settings (we'll create an entry point for external ones)

**Comment by dcramer** [comment]
I will expand with some tests, and work on the identity model likely before merging, though I'm not sure we even need to retain the users token here.

File: src/sentry/integrations/helper.py
Comment (dcramer) [inline src/sentry/integrations/helper.py]: probably should update this when it already exists since this can include access tokens which may have been rotated

**Comment by dcramer** [comment]
I'm going to sort out https://github.com/getsentry/sentry/pull/6005 first I think

**Comment by dcramer** [comment]
I think we're in a good place for merging this as a first pass. Will follow up with more API refactoring/etc.

File: src/sentry/integrations/slack/integration.py
Comment (dcramer) [inline src/sentry/integrations/slack/integration.py]: this likely isn't how the final form of this will look, but still TBD (and will deal w/ it in a follow up patch)

**Comment by dcramer** [comment]
http://i.cra.mr/2305002a393H

**Comment by dcramer** [comment]
I've thrown on some other commits that I'll break out, just didn't wanna split branches this weekend.

**Review by mitsuhiko — DISMISSED** [review]

**Comment by dcramer** [comment]
Added tests for the Slack flow. Will need to have a thumbs up again (we just enabled dismissal of reviews on changes).

**Review by mitsuhiko — DISMISSED** [review]

**Pull request merged**
